### PR TITLE
Feature/diagnostics system port

### PR DESCRIPTION
**Pull request by FejZa**
# XRTK - Mixed Reality Toolkit Change Request

## Overview

Ported the latest implementation of the MRTK DiagnosticsSystem over and made necessary Adjustments. There is some open TODOs:

- [x] `OnDiagnosticsChanged()` in `MixedRealityDiagnosticsSystem.cs` is never executed and the event does not fire respectively. Implementing this will allow to do runtime changes to settings of the diagnostics system, since the Visual Profiler implements `IMixedRealityDiagnosticsSystemHandler`.

- [x] Support for all XRTK platforms Needs to be verified. E.g. while it works on Android mobile AR it actually Displays in 3D space. Probably for mobile s screen space overlay solution would be more appropriate. Also I could not verify Lumin support.
  - [X] Supports UWP
  - [X] Supports Android
  - [X] Supports iOS
  - [x] Supports Lumin

## Target of the change:

- Core (core framework, interfaces and definitions)

## Changes:

- Added new visual profiler component to visualize diagnostics data
- Updated inspector
- Updated profile settings

### CODE REVIEW

**Comment by FejZa** [comment]
XRTK/SDK PR 28 is part of this: https://github.com/XRTK/SDK/pull/28

**Comment by SimonDarksideJ** [comment]
Can we review and update the outstanding tasks prior to merging?
  
Platforms must be tested prior to merge
- Stephen can take care of ML
- I can do HL (when my device returns this week)
- Can you test iOS or do we need help with that?

**Comment by SimonDarksideJ** [comment]
Upi also need to refresh your submodules as they are showing out of date and could cause issue.

**Comment by FejZa** [comment]
Thanks1 I can test iOS later today and check the submodules.

**Comment by StephenHodgson** [comment]
Just a note, make sure the Submodule checkouts are always the latest development checkout sha.

File: XRTK-Core/Packages/com.xrtk.core/Definitions/DiagnosticsSystem/MixedRealityDiagnosticsProfile.cs [hunk 1]
```diff
@@ -2,89 +2,81 @@
...
 using XRTK.Definitions.Utilities;
-using XRTK.Interfaces.Diagnostics;
 
 namespace XRTK.Definitions.Diagnostics
 {
     /// <summary>
     /// Configuration profile settings for setting up diagnostics.
     /// </summary>
-    [CreateAssetMenu(menuName = "Mixed Reality Toolkit/Diagnostics System Profile", fileName = "MixedRealityDiagnosticsProfile", order = (int)CreateProfileMenuItemIndices.Diagnostics)]
+    [CreateAssetMenu(menuName = "Mixed Reality Toolkit/Mixed Reality Diagnostics Profile", fileName = "MixedRealityDiagnosticsProfile", order = (int)CreateProfileMenuItemIndices.Diagnostics)]
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Definitions/DiagnosticsSystem/MixedRealityDiagnosticsProfile.cs]: We can drop the Mixed Reality prefix here I think

File: XRTK-Core/Packages/com.xrtk.core/Definitions/DiagnosticsSystem/MixedRealityDiagnosticsProfile.cs [hunk 2]
```diff
@@ -2,89 +2,81 @@
...
     public class MixedRealityDiagnosticsProfile : BaseMixedRealityProfile
     {
         [SerializeField]
-        [Tooltip("Show diagnostics?")]
-        private bool visible = false;
+        [FormerlySerializedAs("visible")]
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Definitions/DiagnosticsSystem/MixedRealityDiagnosticsProfile.cs]: Nice :)

File: XRTK-Core/Packages/com.xrtk.core/Definitions/DiagnosticsSystem/MixedRealityDiagnosticsProfile.cs (hunk 1 above)
Comment (FejZa) [reply XRTK-Core/Packages/com.xrtk.core/Definitions/DiagnosticsSystem/MixedRealityDiagnosticsProfile.cs]: Will do. Didn't see that one!

File: XRTK-Core/Packages/com.xrtk.core/Services/DiagnosticsSystem/MixedRealityDiagnosticsSystem.cs [hunk 3]
```diff
@@ -1,212 +1,178 @@
...
 using XRTK.Definitions.Diagnostics;
 using XRTK.EventDatum.Diagnostics;
 using XRTK.Interfaces.Diagnostics;
-using Object = UnityEngine.Object;
 
 namespace XRTK.Services.DiagnosticsSystem
 {
     /// <summary>
-    /// The default implementation of the <see cref="IMixedRealityDiagnosticsSystem"/>
+    /// The default implementation of the <see cref="Microsoft.MixedReality.Toolkit.Diagnostics.IMixedRealityDiagnosticsSystem"/>
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Services/DiagnosticsSystem/MixedRealityDiagnosticsSystem.cs]: This doesn't look right

File: XRTK-Core/Packages/com.xrtk.core/Services/DiagnosticsSystem/MixedRealityDiagnosticsSystem.cs [hunk 4]
```diff
@@ -1,212 +1,178 @@
...
-            Visible = MixedRealityToolkit.Instance.ActiveProfile.DiagnosticsSystemProfile.Visible;
-
-            RaiseDiagnosticsChanged();
+            CreateVisualizations();
         }
 
         /// <inheritdoc />
         public override void Destroy()
         {
-            base.Destroy();
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Services/DiagnosticsSystem/MixedRealityDiagnosticsSystem.cs]: We probably want to keep this.

File: XRTK-Core/Packages/com.xrtk.core/Services/DiagnosticsSystem/MixedRealityDiagnosticsSystem.cs [hunk 5]
```diff
@@ -1,212 +1,178 @@
...
-
-            if (Application.isPlaying)
-            {
-                RaiseDiagnosticsChanged();
+                diagnosticVisualizationParent = null;
             }
         }
 
-        #endregion IMixedRealityService Implementation
+        #endregion IMixedRealityService
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Services/DiagnosticsSystem/MixedRealityDiagnosticsSystem.cs]: let's revert this too please.

File: XRTK-Core/Packages/com.xrtk.core/Definitions/DiagnosticsSystem/MixedRealityDiagnosticsProfile.cs (hunk 1 above)
Comment (FejZa) [reply XRTK-Core/Packages/com.xrtk.core/Definitions/DiagnosticsSystem/MixedRealityDiagnosticsProfile.cs]: Updated

File: XRTK-Core/Packages/com.xrtk.core/Services/MixedRealityToolkit.cs [hunk 6]
```diff
@@ -488,7 +488,7 @@ private void InitializeServiceLocator()
 
             if (ActiveProfile.IsDiagnosticsSystemEnabled)
             {
-                if (!CreateAndRegisterService<IMixedRealityDiagnosticsSystem>(ActiveProfile.DiagnosticsSystemSystemType, ActiveProfile.DiagnosticsSystemProfile) || DiagnosticsSystem == null)
+                if (!CreateAndRegisterService<IMixedRealityDiagnosticsSystem>(ActiveProfile.DiagnosticsSystemSystemType, ActiveProfile.DiagnosticsSystemProfile, Instance.MixedRealityPlayspace) || DiagnosticsSystem == null)
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Services/MixedRealityToolkit.cs]: We probably don't need to pass this. Easy enough to grab from the service locator outright

File: XRTK-Core/Packages/com.xrtk.core/Services/DiagnosticsSystem/XRTK.Services.DiagnosticsSystem.asmdef [hunk 7]
```diff
@@ -0,0 +1,14 @@
+{
+    "name": "XRTK.Services.DiagnosticsSystem",
```
Comment (StephenHodgson) [inline XRTK-Core/Packages/com.xrtk.core/Services/DiagnosticsSystem/XRTK.Services.DiagnosticsSystem.asmdef]: Not sure we need an assembly definition for this service. 

What's the benefit of having it?

Comment (FejZa) [reply XRTK-Core/Packages/com.xrtk.core/Services/DiagnosticsSystem/XRTK.Services.DiagnosticsSystem.asmdef]: It was there in MRTK so I kept it. I'll remove it.

**Comment by FejZa** [comment]
Tested on iOS. Behaves the same as on Android.

**Comment by StephenHodgson** [comment]
I'll test it on Lumin soon.

**Comment by StephenHodgson** [comment]
Getting this warning when building to device and running in the editor:
```
A shader supporting instancing could not be found for the VisualProfiler, falling back to traditional rendering. This may impact performance.
```

**Comment by StephenHodgson** [comment]
https://github.com/FejZa/XRTK-Core/pull/1

**Comment by FejZa** [comment]
@StephenHodgson thank you. I'll look into the shader warning. Pretty sure I know what's up.

**Comment by FejZa** [comment]
@StephenHodgson Added missing shader to XRTK.Core/StandardAssets/Shaders. The warning is gone now.

**Comment by StephenHodgson** [comment]
@FejZa just need to resolve the conflicts and we should be good to go.

**Pull request merged**